File: BYTES.NET/WPF/MVVM/ViewModelValidationRule.vb
Namespace WPF.MVVM

    ''' <summary>
    ''' property validation rule class
    ''' </summary>
    ''' <remarks></remarks>
    Public Class ViewModelValidationRule

#Region "public variable(s)"

        Public Enum ResultLevels
            IsNoError = 0
            IsWarning = 1
            IsError = 2
            IsUndefnied = 3
        End Enum

        Public Method As Func(Of Boolean) = Function() False

#End Region

#Region "private variable(s)"

        Private _propertyName As String = Nothing
        Private _outputMessage As String = ""
        Private _resultLevel As ViewModelValidationRule.ResultLevels = ResultLevels.IsError

#End Region

#Region "public properties"

        Public Property PropertyName As String
            Get
                Return _propertyName
            End Get
            Set(value As String)
                _propertyName = value
            End Set
        End Property

        Public Property Message As String
            Get
                Return _outputMessage
            End Get
            Set(value As String)
                _outputMessage = value
            End Set
        End Property

        Public Property Level As ViewModelValidationRule.ResultLevels
            Get
                Return _resultLevel
            End Get
            Set(value As ViewModelValidationRule.ResultLevels)
                _resultLevel = value
            End Set
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' default new instance method
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="validationMethod"></param>
        ''' <param name="message"></param>
        ''' <param name="resultLevel"></param>
        ''' <remarks></remarks>
        Public Sub New(ByVal propertyName As String, ByVal validationMethod As Func(Of Boolean), ByVal message As String, Optional ByVal resultLevel As ViewModelValidationRule.ResultLevels = ResultLevels.IsError)

            'set the properties
            With Me
                .PropertyName = propertyName
                .Method = validationMethod
                .Message = message
                .Level = resultLevel
            End With

        End Sub

#End Region

    End Class

End Namespace

File: src-2023/BYTES.NET.WPF/MVVM/ViewModelValidationRule.cs
//import (default) .net namespace(s) required
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BYTES.NET.WPF.MVVM
{
    /// <summary>
    /// the generic rule type for ViewModel validation
    /// </summary>
    public class ViewModelValidationRule
    {
        #region private variable(s)

        private Func<ViewModelValidationResult[]> _toBeExecutedResult = null;
        private Func<bool?> _toBeExecutedBool = null;

        private string[] _properties = null;
        private string _errorMessage = null;

        #endregion

        #region public properties

        public string[] Properties
        {
            get
            {
                if (_properties == null)
                {
                    return new string[] { };
                }

                return _properties;
            }
        }

        public string ErrorMessage
        {
            get => _errorMessage;
            set => _errorMessage = value;
        }

        #endregion

        #region public new instance method(s)

        /// <summary>
        /// default new instance method
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="properties"></param>
        public ViewModelValidationRule(Func<ViewModelValidationResult[]> toBeExecuted, string[] properties = null)
        {
            _toBeExecutedResult = toBeExecuted;
            _properties = properties;
        }

        /// <summary>
        /// overloaded new instance method, supporting a single property name
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="property"></param>
        public ViewModelValidationRule(Func<ViewModelValidationResult[]> toBeExecuted, string property)
        {
            _toBeExecutedResult = toBeExecuted;
            _properties = new string[] { property };
        }

        /// <summary>
        /// overloaded new instance method, supporting a 'bool' result type delegated function
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="properties"></param>
        public ViewModelValidationRule(Func<bool?> toBeExecuted, string[] properties = null)
        {
            _toBeExecutedBool = toBeExecuted;
            _properties = properties;
        }

        /// <summary>
        /// overloaded new instance method, supporting a 'bool' result type delegated function for a single property
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="property"></param>
        public ViewModelValidationRule(Func<bool?> toBeExecuted, string property)
        {
            _toBeExecutedBool = toBeExecuted;
            _properties = new string[] { property };
        }

        #endregion

        #region public method(s)

        /// <summary>
        /// return the validation result
        /// </summary>
        /// <param name="property"></param>
        /// <returns></returns>
        public ViewModelValidationResult[] Validate(string property = null)
        {
            if (_toBeExecutedResult != null)
            {
                return _toBeExecutedResult();
            }

            bool? result = _toBeExecutedBool();

            if (!result.HasValue)
            {
                return new ViewModelValidationResult[] { new ViewModelValidationResult(ViewModelValidationResult.ResultType.Undefined) };
            }

            if (result == true)
            {

                return new ViewModelValidationResult[] { new ViewModelValidationResult(ViewModelValidationResult.ResultType.Undefined) };

            }

            string message = "Validation failed";

            if (_properties != null)
            {
                if (_properties.Length == 1)
                {
                    message = "Validating '" + _properties[0] + "' failed";
                }
            }

            if (property != null)
            {
                message = "Validating '" + property + "' failed";
            }

            if (_errorMessage != null)
            {
                message = _errorMessage;
            }

            return new ViewModelValidationResult[] { new ViewModelValidationResult(message, ViewModelValidationResult.ResultType.Error) };

        }

        #endregion
    }
}
